openapi.yaml

```diff
@@ -0,0 +1,1252 @@
+        the number of attempts left on the message.
+
+
+        If you wish to fail a message such that it is deleted from the queue and
+        never offered to another consumer, nor placed into the dead letter
+        queue, then call `/queue.complete` instead.
+      operationId: post-v1-queue.defer
+      tags:
+        - Querator V1
+      parameters: []
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/QueueDeferRequest'
+            example:
+              queue_name: queue-name
+              partition: 0
+              items:
+                - id: 2m75RTp9PBx69hw1Q7mjoB0F73Q
+                  enqueue_at: '2024-01-01 12:00:00.0000'
+                  dead: false
+                - id: 2m75RTp9PBx69hw1Q7mjoB0F73R
+                  enqueue_at: '2024-01-01 12:00:00.0000'
+                  dead: false
+                - id: 2m75RTp9PBx69hw1Q7mjoB0F73S
+                  enqueue_at: '2024-01-01 12:00:00.0000'
+                  dead: false
+      responses:
+        '200':
+          description: ''
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/QueueDeferResponse'
+              examples:
+                '1':
+                  summary: Success
+                  value:
+                    code: 200
+          headers: {}
+      security: []
+  /v1/queue.complete:
+    post:
+      summary: Complete reserved items
+      deprecated: false
+      description: >
+        Mark `reserved` items as `completed`. The item can then be removed from
+        the queue by Querator and will not be offered to any other consumers.
+
+
+        If any of the ids provided cannot be marked as `completed` then the
+        entire request will be rejected with a code `453 Request Failed`. It is
+        the responsibility of the client to remove the offending id and try
+        again.
+
+
+        ### Timeout Semantics
+
+        The call will return when the id(s) provided are completed to storage or
+        when `request_timeout` has been reached. If the call returns with a non
+        200 response code, the client should consider the items provided as not
+        completed to the queue. The client should NOT cancel the request before
+        `request_timeout` has been reached. If the call fails to return after
+        `request_timeout` is reached the client can safely cancel the request.
+      operationId: post-v1-queue.complete
+      tags:
+        - Querator V1
+      parameters: []
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/QueueCompleteRequest'
+            example:
+              queue_name: queue-name
+              partition: 0
+              request_timeout: 30s
+              ids:
+                - id-1234
+                - id-1235
+                - id-1236
+      responses:
```

Comment:
If we know the structure - it'd be beneficial to have a 453 payload example. I assume Querator would tell the client which ID(s) failed and why?

Reply:
> I assume Querator would tell the client which ID(s) failed and why?

I've changed the semantics of the API a bit, I wrote an ADR about it. All API calls fail if ANY of the ids or items fail. Querator will never preform a "partial write" or confirmation of anything.